distribution/cachedir/manager.go

```diff
@@ -0,0 +1,190 @@
+package cachedir
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/Sirupsen/logrus"
+)
+
+// Manager handles the creation and deletion of cache directories and has a
+// very particular way of managing the cache. It was designed to be used by a
+// caller that manages the lifetime of the directories, but delegates the
+// mechanics of how to store and locate the directories to the Manager
+type Manager interface {
+	// GetDir creates a new cache dir or returns an existing one if one already
+	// exists for this key from a previous download. It also keeps track of how
+	// many references were made to the cache dir during
+	// ReleaseDir or DeleteDir have to be called exactly once per GetDir call
+	// and GetDir shouldn't be called again until one or the other is called.
+	GetDir(key string) string
+	// ReleaseDir lets the Manager know that the dir is no longer in use but
+	// should be kept around in case we need it again.
+	ReleaseDir(key string)
+	// DeleteDir lets the manager know that the dir is no longer used and can be
+	// deleted. If the delete operation fails and returns an error, the key is
+	// still considered released.
+	DeleteDir(key string) error
+	// CollectGarbage deletes all dirs that are not currently in use (ones that
+	// don't have a GetDir without a matching ReleaseDir/DeleteDir call) and
+	// returns the amount of space saved in bytes.
+	CollectGarbage() (uint64, error)
+}
+
+type manager struct {
+	cacheRoot string
+	mu        sync.Mutex
+	inUse     map[HashedKey]struct{}
+}
+
+// NewManager creates a new cachedir manager
+func NewManager(path string) Manager {
+	return &manager{
+		cacheRoot: path,
+		inUse:     map[HashedKey]struct{}{},
+	}
+}
+
+// HashedKey is a sha256 hash of a cache key which is used to allow using a
+// fixed length but sufficiently unique directory name while allowing "keys" to
+// be of arbitrary length
+type HashedKey string
+
+// hashKey hashes the key to create the directory name because otherwise tche path
+// would be too long. This hash needs to be secure or else multiple downloads could
+// share the same cache directory and cause unexpected and potentially malicious behaviour.
+func hashKey(key string) HashedKey {
+	hash := sha256.New()
+	hash.Write([]byte(key))
+	encoded := make([]byte, hash.Size()*2)
+	hex.Encode(encoded, hash.Sum(nil))
+	return HashedKey(encoded)
+}
+
+func (c *manager) cachePathForKeyHash(keyHash HashedKey) string {
+	return filepath.Join(c.cacheRoot, string(keyHash))
```

Comment:
Note that if you switch to `go-digest`, you'll want to change this to something like:

```go
return filepath.Join(c.cacheRoot, digest.Algorithm(), digest.Hex())
```